### ISSUE TITLE
FluxConcatMap Huge Publish Latency (More than +10x)

### ISSUE THREAD

**Issue by Maithem** [reactor/reactor-core#2558]
## Expected Behavior

I have a simple test to calculate the latency between a producer and a consumer. I have a thread that produces a timestamp every 3 seconds to a queue and a poller that polls the queue every 100ms, if the queue is not empty the timastamp will be published to the subscriber. Once the subscriber receives the timestamp it will subtract the timestamp from the current time thus calculating the propagation delay from the producer to the subscriber. 

Since the poller runs at a frequency of 100ms, I expect the subscriber to observe the values published in less than 200ms.
 

## Actual Behavior

When I run the the test below I see the following output 

```
produced 2021-01-17 16:36:29.205
consumed 2021-01-17 16:36:32.172 2967
produced 2021-01-17 16:36:32.205
consumed 2021-01-17 16:36:34.584 2379
produced 2021-01-17 16:36:35.206
consumed 2021-01-17 16:36:36.994 1788
```

The subscribers observe the published value after 2967, 2379 and 1788 milliseconds respectively. Although the poller sees the produced values right away, it seems like `FluxConcatMap` publishes  to the subscriber after a huge delay  +10x 

## Steps to Reproduce

```kotlin
import reactor.core.publisher.BufferOverflowStrategy
import reactor.core.publisher.Flux
import reactor.core.publisher.SignalType
import reactor.core.scheduler.Schedulers
import java.sql.Timestamp
import java.time.Duration
import java.util.*
import java.util.concurrent.*
import java.util.logging.Level

// Kotlin code
fun main(args: Array<String>) {

    val POLL_INTERVAL: Duration = Duration.ofMillis(100)

    val queue: Queue<Long> = ArrayBlockingQueue(1000);
    val prod = Executors.newSingleThreadExecutor()
    prod.submit {
        while (true) {
            // Periodically produce to the queue every 3 seconds
            val ts = System.currentTimeMillis()
            System.out.println("produced " + Timestamp(ts))
            queue.add(ts)
            Thread.sleep(1000 * 3)
        }
    }

    val scheduler = Schedulers.fromExecutorService(Executors.newSingleThreadScheduledExecutor())

    // Poll the queue at a frequency of 100ms and publish to subscriber
    var publisher: Flux<Long> = Flux.generate<List<Long>> { sink -> sink.next(poll(queue)) }
            .delayElements(POLL_INTERVAL, scheduler)
            .filter { x -> !x.isEmpty() }
            .log("publisher", Level.INFO,
                    SignalType.ON_SUBSCRIBE, SignalType.ON_ERROR, SignalType.CANCEL)
            .concatMap { entry -> Flux.fromIterable(entry) }
            .publish()
            .autoConnect(1)
            .onBackpressureBuffer(10, BufferOverflowStrategy.ERROR)


    publisher.subscribe { entry ->
        run {
            val ts = System.currentTimeMillis()
            // Print the current timestamp and the difference between current time and the timestamp of when the element was produced
            System.out.println("consumed " + Timestamp(ts).toString() + " " + ts.minus(entry).toString())
        }
    }
    Thread.sleep(1000L * 60 * 5)
}

private fun poll(que: Queue<Long>): List<Long> {
    //System.out.println("poll---->" + Timestamp(System.currentTimeMillis()))
    if (!que.isEmpty()) {
        return Arrays.asList(que.poll())
    }
    return Arrays.asList();
}
```

## Possible Solution

I patched the [FluxConcatMap::onSubscribe](https://github.com/reactor/reactor-core/blob/v3.3.4.RELEASE/reactor-core/src/main/java/reactor/core/publisher/FluxConcatMap.java#L203) method from 

```java
int m = f.requestFusion(Fuseable.ANY | Fuseable.THREAD_BARRIER);
```

to 

```java
int m = f.requestFusion(Fuseable.ANY);
```

And the publish latency dropped by 10x as shown below

```
produced 2021-01-17 16:53:46.35
consumed 2021-01-17 16:53:46.537 187
produced 2021-01-17 16:53:49.35
consumed 2021-01-17 16:53:49.543 193
produced 2021-01-17 16:53:52.351
consumed 2021-01-17 16:53:52.456 105
```

It's clear that the latency (i.e., 187ms, 193ms and 105ms)  improved by more than 10x. 

It is my understanding that `Fuseable.THREAD_BARRIER` is required when the publish happens on a different scheduler/executor, but in this case it's not, the subscribers call backs are executed on the same publisher's executor. 
I'm not very familiar with this code base and haven't spent too much time studying it, so it's possible that a more elaborate fix is required to handle both cases where publish happens on the same and different executor. 

## Your Environment

* Reactor version(s) used: 3.3.4.RELEASE
* JVM version (`java -version`): Java HotSpot(TM) 64-Bit Server VM (build 25.201-b09, mixed mode)
* OS and version (eg `uname -a`): 4.15.0-128-generic #131~16.04.1-Ubuntu

**Comment by simonbasle** [comment]
Note that `3.3.4.RELEASE` is 9 versions behind (https://github.com/reactor/reactor-core/releases/tag/v3.3.13.RELEASE). Not saying this would automatically improve the situation, but it can't hurt to try with the latest 3.3.x release...

That said, from briefly looking at your reproduction code, I'm puzzled by what you're trying to achieve:
 - you're using a blocking queue
 - you're delaying elements that come out of the queue, which to me sounds like it would naturally compound the latency
 - you say there's no thread boundary, yet `delayElements` does introduce such a boundary (runs on `Schedulers.parallel()` by default)...

**Comment by Maithem** [comment]
> but it can't hurt to try with the latest 3.3.x release

I tried with the latest and it doesn't help. The problem seems to have been introduced post `3.1.0.RELEASE`.

> you're using a blocking queue

Although it is a blocking queue the consume part isn't (i.e., `Queue::poll` doesn't block)

> you're delaying elements that come out of the queue, which to me sounds like it would naturally compound the latency

From the documentation [delayElements](https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/publisher/Flux.java#L3967) it seems like empty sequences shouldn't be delayed. 


Basically the code is doing the following:

1.  Create a producer thread that will write to the queue every 3 seconds
2.  Create a poller thread that will poll the queue  at a **fixed** frequency of  **100ms** via flux and call the subscribers once it observes new elements from (1). Notice that the flux subscribers are only interested in non-empty sequences.


My expectation is that once the producer thread writes to the queue, the flux subscribers should be notified right away.

**Comment by simonbasle** [comment]
since `delayElements` compounds with the latency of the `onNext` themselves, have you tried a different approach for you "poller thread"? For instance, `Flux.interval` would sound like a preferable approach for a truly fixed frequency. Then you can use `handle` to poll the queue and avoid any signal in case the queue is empty:

```java
Flux.interval(Duration.ofMillis(100), scheduler)
    .handle((tick, sink) -> {
        List<Long> polled = poll(queue);
        if (polled != null) sink.next(polled);
    });
```

**Comment by Maithem** [comment]
Thanks for the suggestion, I didn't know about that API.

I used a different approach to solve my problem. Limiting the generate with `.limitRate(1)` seems to also do the trick, but its not as accurate as the method you suggested since. It seems like the generate is producing 1-2 task at a time despite the limit being one. 

```kotlin

    var publisher: Flux<Long> = Flux.generate<List<Long>> { sink -> sink.next(poll(queue)) }
            .limitRate(1)
            .delayElements(POLL_INTERVAL, scheduler)
            .filter { x -> !x.isEmpty() }
            .log("publisher", Level.INFO,
                    SignalType.ON_SUBSCRIBE, SignalType.ON_ERROR, SignalType.CANCEL)
            .concatMap { entry -> Flux.fromIterable(entry) }
            .publish()
            .autoConnect(1)
            .onBackpressureBuffer(10, BufferOverflowStrategy.ERROR)

```

**Comment by simonbasle** [comment]
Closing as it seems multiple alternatives have been found.
